vllm/model_executor/layers/fused_moe/layer.py

```diff
@@ -210,92 +198,99 @@ def weight_loader(self, param: torch.nn.Parameter,
         # Weight scales
         elif "weight_scale" in weight_name:
             # If we are in merged column case (gate_up_proj)
-            #   shard_id 0 == gate_proj / w1
-            #   shard_id 2 == up_proj / w3
-            if shard_id == 0 or shard_id == 2:
+            if shard_id == "w1" or shard_id == "w3":
                 # We have to keep the weight scales of w1 and w3 because
                 # we need to re-quantize w1/w3 weights after weight loading.
-                idx = 0 if shard_id == 0 else 1
+                idx = 0 if shard_id == "w1" else 1
                 param_data[expert_id][idx] = loaded_weight
             # If we are in the row parallel case (down_proj)
-            #   shard_id 1 == down_proj / w2
             else:
                 param_data[expert_id] = loaded_weight
-        # Weights
+
+    def weight_loader(self, param: torch.nn.Parameter,
+                      loaded_weight: torch.Tensor, weight_name: str,
+                      shard_id: str, expert_id: int) -> None:
+        if shard_id not in ["w1", "w2", "w3"]:
+            raise ValueError(f"shard_id must be ['w1','w2','w3'] but "
+                             f"got {shard_id}.")
+
+        # Special case for fp8 scales.
+        if getattr(param, "is_fp8_scale", False):
+            self._load_fp8_scale(param.data, loaded_weight, weight_name,
+                                 shard_id, expert_id)
+            return
+
+        expert_data = param.data[expert_id]
+        tp_rank = get_tensor_model_parallel_rank()
+
+        # If transposed, weight is saved as [input_dim, output_dim]
+        # Otherwise, weight is saved as     [output_dim, input_dim]
+        is_transposed = getattr(param, "is_transposed", False)
+        input_dim = 0 if is_transposed else 1
+        output_dim = 1 if is_transposed else 0
```

Comment:
We should read input_dim and output_dim from the layer instead, this is a common pattern in the codebase: https://github.com/vllm-project/vllm/blob/05308891e203329a733bcf29a3452b15b75b5eb4/vllm/model_executor/layers/quantization/fp8.py#L151-L155